### PR TITLE
Python: change bootstrapping.

### PR DESCRIPTION
**Pull request by FRidh**
## Description of changes

This is https://github.com/NixOS/nixpkgs/pull/245509 but then on `python-updates` branch so we can use Hydra.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tjni** [comment]
I have rebased this on staging-next for now to make it a little easier to reason about for me. I propose:

1. For commits that are merged into `master`, pick it up after it's been merged into `staging-next`.
2. For commits that are merged into `staging`, cherry-pick over to here and when it comes time to rebase this onto `staging`, it should be de-duplicated.

**Comment by FRidh** [comment]
I'm going to start one more run. I'm pretty confident we can merge then. Please don't add any more unrelated updates, only (larger) fixes.

**Comment by FRidh** [comment]
jedi has one failing test on all platforms which has to be resolved

**Comment by tjni** [comment]
`jedi` needed to be updated to 0.19.0 to be compatible with `attrs` 23.1.0

**Comment by tjni** [comment]
The last 3 commits are PRs that were either merged or are targeting `master` but they should unblock testing of a few hundred dependent packages all together.

**Comment by FRidh** [comment]
`pg8000` will need to be fixed
```
adding license file 'LICENSE'
writing manifest file 'pg8000.egg-info/SOURCES.txt'

ERROR Missing dependencies:
        versioningit>=2.1.0
```
After that we can trigger hydra again.

**Comment by doronbehar** [comment]
Small scipy update at #249921 .

**Comment by FRidh** [comment]
About 4000 failures on this branch.

**Comment by tjni** [comment]
I looked through a good number of the failures and listed them at the top of https://github.com/NixOS/nixpkgs/issues/247287. The biggest improvement will be from fixing `home-assistant` packages (https://github.com/NixOS/nixpkgs/pull/250124). I'm not entirely sure where else we can get big improvements. I'd like to merge the outstanding PRs, merge those changes in `master` into `staging-next`, and then rebase on top again to get an up to date list of broken packages.

File: pkgs/development/python-modules/pip/default.nix [hunk 1]
```diff
@@ -14,40 +16,46 @@
...
   nativeCheckInputs = [ mock scripttest virtualenv pretend pytest ];
+
   # Pip wants pytest, but tests are not distributed
   doCheck = false;
 
+  postInstall = ''
+    installShellCompletion --cmd pip \
+      --bash <($out/bin/pip completion --bash) \
+      --fish <($out/bin/pip completion --fish) \
+      --zsh <($out/bin/pip completion --zsh)
```
Comment (anund) [inline pkgs/development/python-modules/pip/default.nix:50]: `<($out/bin/pip completion --zsh --no-cache-dir)` will properly clear the warning about `.cache/pip` not being writable. zsh may not work until https://github.com/pypa/pip/issues/12166 is closed. (given pip is at the bottom of ~40000 packages leaving it in seems fine)

Comment (anund) [reply pkgs/development/python-modules/pip/default.nix:50]: Also fixes https://github.com/NixOS/nixpkgs/issues/224610

**Comment by FRidh** [comment]
Now with home-assistant fixed this is good to go. The items on the list https://github.com/NixOS/nixpkgs/issues/247287 can be done with individual commits, while this change is on staging/staging-next and even on master afterwards by the individual maintainers if needed.

File: pkgs/development/python-modules/pip/default.nix (hunk 1 above)
Comment (tjni) [reply pkgs/development/python-modules/pip/default.nix:50]: Thank you for figuring that out. This was merged without that change, but we should add this in a separate commit.

Comment (anund) [reply pkgs/development/python-modules/pip/default.nix:50]: Please do. I'm working on another change atm and don't need the two things coupled.

Comment (anund) [reply pkgs/development/python-modules/pip/default.nix:50]: Tacked it on to this change https://github.com/NixOS/nixpkgs/pull/250398/files

Comment (tjni) [reply pkgs/development/python-modules/pip/default.nix:50]: Thank you, I appreciate you creating that PR. I'm sorry I hadn't found the time yet today.

File: pkgs/development/python-modules/scramp/default.nix [hunk 2]
```diff
@@ -40,6 +42,7 @@ buildPythonPackage rec {
 
   postPatch = ''
     # Upstream uses versioningit to set the version
+    sed -i "/versioningit >=/d" pyproject.toml
     sed -i '/^name =.*/a version = "${version}"' pyproject.toml
     sed -i "/dynamic =/d" pyproject.toml
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/scramp/default.nix:47]: nit: those could have been combined into one command

Comment (tjni) [reply pkgs/development/python-modules/scramp/default.nix:47]: Yes, you're right. At the same time, even though it's less efficient, I feel that keeping separate commands here is more readable. But this motivates me to look into whether we can handle versioningit better in general, thank you!

Comment (SuperSandro2000) [reply pkgs/development/python-modules/scramp/default.nix:47]: I always find less code overhault more readable unless you are using some special character shortcuts. Like here, eg deduplicating pyproject.toml but as said it is not that super important and focusing on getting a setup-hook for verioningit is a better solution anyway.

**Comment by jqqqqqqqqqq** [comment]
Should we backport this to 23.05?

I'm seeing errors on 23.05 because some inconsistency happened:

```
[ERROR]   stderr) error: anonymous function at /nix/store/r5b4xjc4gvraxdslr98f1qpn62vp3gw8-source/pkgs/development/python-modules/wheel/default.nix:1:1 called with unexpected argument 'flit-core'
[ERROR]   stderr)
[ERROR]   stderr)        at /nix/store/r5b4xjc4gvraxdslr98f1qpn62vp3gw8-source/lib/customisation.nix:80:16:
[ERROR]   stderr)
[ERROR]   stderr)            79|     let
[ERROR]   stderr)            80|       result = f origArgs;
[ERROR]   stderr)              |                ^
[ERROR]   stderr)            81|
[ERROR]   stderr) (use '--show-trace' to show detailed location information)
[ERROR]  failure) Child process exited with error code: 1
```

And 23.05 is still using the old wheel package:

```
cat /nix/store/r5b4xjc4gvraxdslr98f1qpn62vp3gw8-source/pkgs/development/python-modules/wheel/default.nix
{ lib
, buildPythonPackage
, fetchFromGitHub
, bootstrapped-pip
, setuptools
}:

buildPythonPackage rec {
  pname = "wheel";
  version = "0.38.4";
  format = "other";

  src = fetchFromGitHub {
    owner = "pypa";
    repo = pname;
    rev = version;
    hash = "sha256-yZLU0t/nz6kfnnoLL15bybOxN4+SJUaTJsCpGffl1QU=";
    name = "${pname}-${version}-source";
    postFetch = ''
...
```

**Comment by kirillrdy** [comment]
> Should we backport this to 23.05?
> 
> I'm seeing errors on 23.05 because some inconsistency happened:
> 
> ```
> [ERROR]   stderr) error: anonymous function at /nix/store/r5b4xjc4gvraxdslr98f1qpn62vp3gw8-source/pkgs/development/python-modules/wheel/default.nix:1:1 called with unexpected argument 'flit-core'
> [ERROR]   stderr)
> [ERROR]   stderr)        at /nix/store/r5b4xjc4gvraxdslr98f1qpn62vp3gw8-source/lib/customisation.nix:80:16:
> [ERROR]   stderr)
> [ERROR]   stderr)            79|     let
> [ERROR]   stderr)            80|       result = f origArgs;
> [ERROR]   stderr)              |                ^
> [ERROR]   stderr)            81|
> [ERROR]   stderr) (use '--show-trace' to show detailed location information)
> [ERROR]  failure) Child process exited with error code: 1
> ```
> 
> And 23.05 is still using the old wheel package:
> 
> ```
> cat /nix/store/r5b4xjc4gvraxdslr98f1qpn62vp3gw8-source/pkgs/development/python-modules/wheel/default.nix
> { lib
> , buildPythonPackage
> , fetchFromGitHub
> , bootstrapped-pip
> , setuptools
> }:
> 
> buildPythonPackage rec {
>   pname = "wheel";
>   version = "0.38.4";
>   format = "other";
> 
>   src = fetchFromGitHub {
>     owner = "pypa";
>     repo = pname;
>     rev = version;
>     hash = "sha256-yZLU0t/nz6kfnnoLL15bybOxN4+SJUaTJsCpGffl1QU=";
>     name = "${pname}-${version}-source";
>     postFetch = ''
> ...
> ```

I don't think it fits backporting criteria, also 23.11 is imminent

**Pull request merged**